File: code-example-tests/javascript/driver/tests/vector_search/local-rag.test.js
/**
 * Tests for the local RAG examples (local embedding model + local LLM).
 *
 * These tests run against real services — nothing is mocked:
 *
 * - Embeddings come from the real `mixedbread-ai/mxbai-embed-large-v1` model
 *   through `@xenova/transformers`, downloaded from the Hugging Face model hub
 *   on first run and cached locally afterward.
 * - Retrieval runs a real `$vectorSearch` query against the
 *   `sample_airbnb.listingsAndReviews` collection on the deployment in
 *   CONNECTION_STRING. The deployment must have Search and Vector Search
 *   installed (a local Atlas deployment or an Atlas cluster).
 * - Generation runs the real Mistral 7B model locally through `gpt4all`. The
 *   model file is ~4 GB and downloads to the gpt4all cache on first run.
 *
 * Sample data hygiene: these tests add an `embeddings` field to documents in
 * `sample_airbnb.listingsAndReviews` and create a `vector_index` search index
 * on that collection. Both are reverted in teardown. The sample database
 * itself is never dropped.
 *
 * Because the two models together are several gigabytes and download on first
 * run, these tests are opt-in. Set `RUN_LOCAL_MODEL_TESTS=true` to run them.
 * `@xenova/transformers` and `gpt4all` are optional dependencies installed
 * separately (see package-models.json); CI leaves the variable unset, so the
 * suite skips there and never needs them.
 */

import { MongoClient } from 'mongodb';
import Expect from '../../utils/Expect.js';
import { checkSampleDataAvailable } from '../../utils/sampleDataChecker.js';

const DATABASE_NAME = 'sample_airbnb';
const COLLECTION_NAME = 'listingsAndReviews';
const INDEX_NAME = 'vector_index';
const EMBEDDING_DIMENSIONS = 1024;

// Model downloads and a 50-document embedding pass are slow; allow generous
// time so a first run that fetches the models does not fail on the timeout.
const MODEL_TIMEOUT_MS = 30 * 60 * 1000;

function runLocalModelTests() {
  return ['1', 'true', 'yes'].includes(
    (process.env.RUN_LOCAL_MODEL_TESTS || '').toLowerCase()
  );
}

const describeOrSkip = runLocalModelTests() ? describe : describe.skip;

describeOrSkip('Vector Search local RAG tutorial', () => {
  let client;
  let collection;
  let sampleDataAvailable = false;
  let createdIndex = false;
  let generateEmbeddings;
  let createVectorIndex;
  let runTestQuery;
  let answerQuestion;

  beforeAll(async () => {
    if (!process.env.CONNECTION_STRING) {
      throw new Error(
        'Could not retrieve CONNECTION_STRING - make sure you have created ' +
          'the .env file at the root of the javascript/driver directory and ' +
          'the variable is correctly named as CONNECTION_STRING.'
      );
    }

    sampleDataAvailable = await checkSampleDataAvailable(DATABASE_NAME, [
      COLLECTION_NAME,
    ]);

    client = new MongoClient(process.env.CONNECTION_STRING);
    await client.connect();
    collection = client.db(DATABASE_NAME).collection(COLLECTION_NAME);

    ({ generateEmbeddings } =
      await import('../../examples/vector_search/local_rag/generate-embeddings.js'));
    ({ createVectorIndex } =
      await import('../../examples/vector_search/local_rag/vector-index.js'));
    ({ runTestQuery } =
      await import('../../examples/vector_search/local_rag/test-query.js'));
    ({ answerQuestion } =
      await import('../../examples/vector_search/local_rag/local-llm.js'));
  }, MODEL_TIMEOUT_MS);

  afterAll(async () => {
    if (!client) return;
    await removeEmbeddings();
    if (createdIndex) {
      try {
        await collection.dropSearchIndex(INDEX_NAME);
      } catch {
        // The index may already be gone; teardown should not fail the run.
      }
    }
    await client.close();
  });

  // Revert embeddings between tests so each one starts from the collection's
  // original state rather than inheriting a prior test's writes.
  beforeEach(async () => {
    await removeEmbeddings();
  });

  async function removeEmbeddings() {
    if (!collection) return;
    await collection.updateMany(
      { embeddings: { $exists: true } },
      { $unset: { embeddings: '' } }
    );
  }

  /** Creates the vector index once, using the example under test. */
  async function ensureIndex() {
    const existing = await collection.listSearchIndexes().toArray();
    if (!existing.some((index) => index.name === INDEX_NAME)) {
      await createVectorIndex();
      createdIndex = true;
    }
  }

  async function waitForQueryableIndex(timeoutSeconds = 300) {
    const deadline = Date.now() + timeoutSeconds * 1000;
    while (Date.now() < deadline) {
      const indexes = await collection.listSearchIndexes(INDEX_NAME).toArray();
      if (indexes.length && indexes[0].queryable === true) return;
      await new Promise((resolve) => setTimeout(resolve, 5000));
    }
    throw new Error(
      `Search index ${INDEX_NAME} did not become queryable within ` +
        `${timeoutSeconds} seconds.`
    );
  }

  function skipWithoutSampleData() {
    if (!sampleDataAvailable) {
      console.warn(
        `\n⚠️  Skipping - Missing sample data: ${DATABASE_NAME}.${COLLECTION_NAME}`
      );
      return true;
    }
    return false;
  }

  describe('generate-embeddings.js', () => {
    it(
      'Should report fifty updated documents',
      async () => {
        if (skipWithoutSampleData()) return;

        const result = await generateEmbeddings();

        Expect.that(result).shouldMatch(
          'vector_search/local_rag/local-rag-create-embeddings-output.sh'
        );
      },
      MODEL_TIMEOUT_MS
    );

    it(
      'Should store a 1024-dimension vector on each updated document',
      async () => {
        if (skipWithoutSampleData()) return;

        await generateEmbeddings();

        const document = await collection.findOne(
          { embeddings: { $exists: true } },
          { projection: { embeddings: 1, _id: 0 } }
        );

        Expect.that({ dimensions: document.embeddings.length }).shouldMatch({
          dimensions: EMBEDDING_DIMENSIONS,
        });
      },
      MODEL_TIMEOUT_MS
    );
  });

  describe('vector-index.js', () => {
    it(
      'Should create a queryable vectorSearch index',
      async () => {
        if (skipWithoutSampleData()) return;

        await ensureIndex();
        await waitForQueryableIndex();

        const indexes = await collection
          .listSearchIndexes(INDEX_NAME)
          .toArray();

        Expect.that({
          name: indexes[0].name,
          type: indexes[0].type,
        }).shouldMatch({ name: INDEX_NAME, type: 'vectorSearch' });
      },
      MODEL_TIMEOUT_MS
    );
  });

  describe('retrieve-documents.js', () => {
    it(
      'Should return five listings with summary, listing_url, and descending scores',
      async () => {
        if (skipWithoutSampleData()) return;

        await generateEmbeddings();
        await ensureIndex();
        await waitForQueryableIndex();

        const results = await runTestQuery();

        // Vector search scores and result order vary across environments, so
        // validate the shape of the results rather than exact values.
        Expect.that(results)
          .shouldResemble(
            'vector_search/local_rag/local-rag-query-results-output.sh'
          )
          .withSchema({
            count: 5,
            requiredFields: ['summary', 'listing_url', 'score'],
          });

        // The index uses cosine similarity, so each score must be a number in
        // the normalized (0, 1] range, and $vectorSearch must return the
        // results from most to least relevant.
        const scores = results.map((result) => result.score);

        Expect.that({
          scoresInRange: scores.every(
            (score) => Number.isFinite(score) && score > 0 && score <= 1
          ),
          scoresDescending: scores.every(
            (score, index) => index === 0 || scores[index - 1] >= score
          ),
        }).shouldMatch({ scoresInRange: true, scoresDescending: true });
      },
      MODEL_TIMEOUT_MS
    );
  });

  describe('local-llm.js', () => {
    it(
      'Should return a generated answer from the retrieved listings',
      async () => {
        if (skipWithoutSampleData()) return;

        await generateEmbeddings();
        await ensureIndex();
        await waitForQueryableIndex();

        const answer = await answerQuestion();

        // The LLM response text varies on every run, so assert only that the
        // generation step produced text.
        Expect.that({
          answer,
          isEmpty: answer.trim() === '',
        }).shouldMatch({ answer: '...', isEmpty: false });
      },
      MODEL_TIMEOUT_MS
    );
  });
});

File: code-example-tests/go/driver/tests/vector_search/local_rag_test.go
package vector_search

import (
	"context"
	"net"
	"os"
	"strings"
	"testing"
	"time"

	"driver-examples/examples/vector_search/local_rag"
	"driver-examples/utils"
	"driver-examples/utils/compare"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

const (
	databaseName   = "sample_airbnb"
	collectionName = "listingsAndReviews"
)

// queryResult mirrors the fields the tutorial prints for each matching
// document, so schema validation can check the shape of the search results.
type queryResult struct {
	Summary    string  `bson:"summary"`
	ListingURL string  `bson:"listing_url"`
	Score      float32 `bson:"score"`
}

// requiresOllama skips the test when a local Ollama server is unreachable. The
// examples generate embeddings with nomic-embed-text and answers with mistral,
// both served by Ollama on the host in OLLAMA_HOST (or localhost:11434).
func requiresOllama(t *testing.T) {
	t.Helper()
	utils.EnsureEnvLoaded()

	address := os.Getenv("OLLAMA_HOST")
	if address == "" {
		address = "localhost:11434"
	}
	address = strings.TrimPrefix(strings.TrimPrefix(address, "http://"), "https://")
	address = strings.TrimSuffix(address, "/")

	conn, err := net.DialTimeout("tcp", address, 5*time.Second)
	if err != nil {
		t.Skipf("skipping: no Ollama server reachable at %s: %v", address, err)
	}
	_ = conn.Close()
}

// setupTestDB connects to the deployment and returns a cleanup function that
// reverts every change the examples make to the sample data. The sample
// database itself is never dropped.
func setupTestDB(t *testing.T) (*mongo.Client, func()) {
	t.Helper()
	ctx := context.Background()

	uri := utils.GetConnectionString()
	if uri == "" {
		t.Fatal("set your 'CONNECTION_STRING' environment variable")
	}

	client, err := mongo.Connect(options.Client().ApplyURI(uri))
	if err != nil {
		t.Fatalf("failed to connect to the server: %v", err)
	}

	cleanup := func() {
		coll := client.Database(databaseName).Collection(collectionName)

		unset := bson.D{{Key: "$unset", Value: bson.D{
			{Key: local_rag.EmbeddingsField, Value: ""},
			{Key: "pageContent", Value: ""},
			{Key: "metadata", Value: ""},
		}}}
		filter := bson.D{{Key: local_rag.EmbeddingsField, Value: bson.D{{Key: "$exists", Value: true}}}}
		if _, err := coll.UpdateMany(ctx, filter, unset); err != nil {
			t.Logf("failed to remove embeddings: %v", err)
		}

		if err := coll.SearchIndexes().DropOne(ctx, local_rag.VectorIndexName); err != nil {
			t.Logf("failed to drop search index: %v", err)
		}

		if err := client.Disconnect(ctx); err != nil {
			t.Logf("failed to disconnect client: %v", err)
		}
	}

	return client, cleanup
}

// TestLocalRag runs the local RAG tutorial end to end. The steps are ordered
// subtests of a single test because each step depends on the previous one: the
// index needs embeddings, and retrieval needs a queryable index.
func TestLocalRag(t *testing.T) {
	utils.RequiresSampleDataWithCollections(t, map[string][]string{
		databaseName: {collectionName},
	})
	requiresOllama(t)

	client, cleanup := setupTestDB(t)
	defer cleanup()

	t.Run("GenerateEmbeddings", func(t *testing.T) {
		result := local_rag.GenerateEmbeddings()
		compare.ExpectThat(t, result).
			ShouldMatch("examples/vector_search/local_rag/generate-embeddings-output.txt")
	})

	t.Run("EmbeddingsHave768Dimensions", func(t *testing.T) {
		coll := client.Database(databaseName).Collection(collectionName)
		filter := bson.D{{Key: local_rag.EmbeddingsField, Value: bson.D{{Key: "$exists", Value: true}}}}

		var listing local_rag.Listing
		if err := coll.FindOne(context.Background(), filter).Decode(&listing); err != nil {
			t.Fatalf("failed to find a document with embeddings: %v", err)
		}

		actual := []bson.D{{{Key: "dimensions", Value: len(listing.Embeddings)}}}
		expected := []bson.D{{{Key: "dimensions", Value: 768}}}
		compare.ExpectThat(t, actual).ShouldMatch(expected)
	})

	t.Run("CreateVectorIndex", func(t *testing.T) {
		indexName := local_rag.CreateVectorIndex()

		actual := []bson.D{{{Key: "name", Value: indexName}}}
		expected := []bson.D{{{Key: "name", Value: local_rag.VectorIndexName}}}
		compare.ExpectThat(t, actual).ShouldMatch(expected)
	})

	t.Run("RetrieveDocuments", func(t *testing.T) {
		documents := local_rag.RetrieveDocuments("beach house")

		results := make([]queryResult, len(documents))
		for i, document := range documents {
			listingURL, ok := document.Metadata["listing_url"].(string)
			if !ok {
				t.Fatalf("expected a string listing_url in the metadata of document %d", i)
			}
			results[i] = queryResult{
				Summary:    document.PageContent,
				ListingURL: listingURL,
				Score:      document.Score,
			}
		}

		// Vector search scores and result order vary across deployments, so
		// validate the shape of the results rather than exact values.
		compare.ExpectThat(t, results).
			ShouldResemble("examples/vector_search/local_rag/retrieve-documents-output.txt").
			WithSchema(compare.Schema{
				Count:          5,
				RequiredFields: []string{"summary", "listing_url", "score"},
			})
	})

	t.Run("TestQueryBuildsContext", func(t *testing.T) {
		queryContext := local_rag.RunTestQuery()

		actual := []bson.D{{
			{Key: "context", Value: queryContext},
			{Key: "is_empty", Value: strings.TrimSpace(queryContext) == ""},
		}}
		compare.ExpectThat(t, actual).ShouldMatch(`[{"context": "...", "is_empty": false}]`)
	})

	t.Run("AnswerQuestion", func(t *testing.T) {
		answer := local_rag.AnswerQuestion()

		// The LLM response text varies on every run, so assert only that the
		// generation step produced text.
		actual := []bson.D{{
			{Key: "answer", Value: answer},
			{Key: "is_empty", Value: strings.TrimSpace(answer) == ""},
		}}
		compare.ExpectThat(t, actual).ShouldMatch(`[{"answer": "...", "is_empty": false}]`)
	})
}
